cluster-autoscaler/processors/nodeinfos/podtemplates/node_info_with_podtemplates_processors.go

```diff
@@ -0,0 +1,172 @@
+		// If not a NodeTemplate, Pods from PodTemplates should alreadt be already present
+		// in the PodList attached to the Node.
+		if strings.HasPrefix(nodeInfo.Node().Name, "template-node-for-") {
+			var err error
+			newNodeInfo, err = getNodeInfoWithPodTemplates(nodeInfo, podTemplates, clusterSnapshot, ctx.PredicateChecker)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		} else {
+			newNodeInfo = nodeInfosForNodeGroups[id]
+		}
+
+		result[id] = newNodeInfo
+	}
+
+	return result, utilerrors.NewAggregate(errs)
+}
+
+// CleanUp cleans up processor's internal structuxres.
+func (p *nodeInfoWithPodTemplateProcessor) CleanUp() {
+	p.cancelFunc()
+}
+
+func newPodTemplateLister(kubeClient client.Interface, stopchannel <-chan struct{}) v1lister.PodTemplateLister {
+	podTemplateWatchOption := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.Everything().String()
+		options.LabelSelector = labels.SelectorFromSet(getDaemonsetPodTemplateLabelSet()).String()
+	}
+	listWatcher := cache.NewFilteredListWatchFromClient(kubeClient.CoreV1().RESTClient(), "podtemplates", apiv1.NamespaceAll, podTemplateWatchOption)
+	store, reflector := cache.NewNamespaceKeyedIndexerAndReflector(listWatcher, &apiv1.PodTemplate{}, time.Hour)
+	lister := v1lister.NewPodTemplateLister(store)
+	go reflector.Run(stopchannel)
+	return lister
+}
+
+func getNodeInfoWithPodTemplates(baseNodeInfo *schedulerframework.NodeInfo, podTemplates []*apiv1.PodTemplate, clusterSnapshot *simulator.BasicClusterSnapshot, predicateChecker simulator.PredicateChecker) (*schedulerframework.NodeInfo, error) {
+	node := baseNodeInfo.Node()
+	var pods []*apiv1.Pod
+
+	for _, podInfo := range baseNodeInfo.Pods {
+		pods = append(pods, podInfo.Pod)
+	}
+	if err := clusterSnapshot.AddNodeWithPods(node, pods); err != nil {
+		return nil, err
+	}
+	// clone baseNodeInfo to not modify the input object.
+	newNodeInfo := baseNodeInfo.Clone()
+	for _, podTpl := range podTemplates {
+		newPod := newPod(podTpl, node.Name)
+		err := predicateChecker.CheckPredicates(clusterSnapshot, newPod, node.Name)
+		if err == nil {
+			newNodeInfo.AddPod(newPod)
+		} else if err.ErrorType() == simulator.NotSchedulablePredicateError {
+			// ok; we are just skipping this daemonset
+		} else {
+			// unexpected error
+			return nil, fmt.Errorf("unexpected error while calling PredicateChecker; %v", err)
+		}
+	}
+
+	return newNodeInfo, nil
+}
+
+func getDaemonsetPodTemplateLabelSet() labels.Set {
+	daemonsetPodTemplateLabels := map[string]string{
+		PodTemplateDaemonSetLabelKey: PodTemplateDaemonSetLabelValueTrue,
+	}
+	return labels.Set(daemonsetPodTemplateLabels)
+}
+
+func newPod(pt *apiv1.PodTemplate, nodeName string) *apiv1.Pod {
+	newPod := &apiv1.Pod{Spec: pt.Template.Spec, ObjectMeta: pt.Template.ObjectMeta}
+	newPod.Namespace = pt.Namespace
+	newPod.Name = fmt.Sprintf("%s-pod", pt.Name)
+	newPod.Spec.NodeName = nodeName
```

Comment:
Setting nodeName means the pod is already scheduled on a given node. Here you want to check if it would be able to schedule, so I don't think it makes sense to set this field.

Reply:
Here I want to simulate a Pod that was/will be create by a Daemonset, it is why I force the scheduling on this specific node by setting the NodeName. I used the same logic used in `GetDaemonSetPodsForNode` with the `newPod` function ([link](https://github.com/kubernetes/autoscaler/blob/a839343bbd737534133d110e7decf295f8281baa/cluster-autoscaler/utils/daemonset/daemonset.go#L68))

But it is true that I can use the  `pod.spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution` to fully simulate what the daemonset-controller is doing.